singularity-eos/eos/eos_gruneisen.cpp

```diff
@@ -12,12 +12,155 @@
+        root2 = (-4 * s3 * s2 * s1 - 9 * square(s3) + cube(s2)) /
+                (-s3 * (square(s2) - 3 * s3 * s1));
+        root = find_min_bounded_val(root1, root2, 0, 1);
+      }
+    } else {
+      // We'll need an iterative method to search for the minimum bounded root
+      Real minbound = 0;
+      Real maxbound = 1;
+      if (discriminant > 0) {
+        // Three real roots. We need to use the extrema to ensure we have a proper bracket
+        // in which to search for the root.
+        const Real extremum1 =
+            (2 * s2 + std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real extremum2 =
+            (2 * s2 - std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real min_extremum = std::min(extremum1, extremum2);
+        const Real max_extremum = std::max(extremum1, extremum2);
+        if (s3 < 0) {
+          // Cubic is *increasing*
+          // Because poly(eta = 0) = 1, the only possible root for an increasing function
+          // will lie between the etrema.
+          minbound = std::max(min_extremum, minbound);
+          maxbound = std::min(std::max(max_extremum, minbound), maxbound);
+        } else {
+          // Cubic is *decreasing*
+          // Because poly(eta = 0) = 1, the only possible root for a decreasing function
+          // will lie outside of the extrema. Further, the only possibility for multiple
+          // bound roots will occur when the extremum are both positive
+          if (min_extremum > 0) {
+            maxbound = std::min(min_extremum, maxbound);
+          }
+        }
+      }
+      if (poly(minbound) * poly(maxbound) <= 0) {
+        // Root is appropriately bounded
+        using RootFinding1D::bisect;
+        using RootFinding1D::RootCounts;
+        using RootFinding1D::Status;
+        RootCounts counts;
+        const Real xtol = 1.e-08;
+        const Real ytol = 1.e-08;
+        const Real eta_guess = 0.001;
+        auto status =
+            bisect(poly, 0., eta_guess, minbound, maxbound, xtol, ytol, root, counts);
+        if (status != Status::SUCCESS) {
+          // Root finder failed even though the solution was bracketed... this is an error
+          std::stringstream errorMessage;
+          errorMessage << "Gruneisen initialization: Cubic root find failed. Maximum "
+                          "density cannot be"
+                       << " automatically determined and must be manually specified."
+                       << std::endl;
+          EOS_ERROR(errorMessage.str().c_str());
+        }
+      } else {
+        // Root doesn't lie within physical bounds for eta so no maximum density exists
+        root = -1.;
+      }
+    } // Cubic discriminant >= 0
+  }   // Linear/Quadratic/Cubic
+  if (root > 0) {
+    return rho0 / (1 - root);
+  } else {
+    // No bounded root exists so there is no upper-limit on the compression
+    return 1.e99;
```

Comment:
should you return infinity here? Or will that cause problems?

Reply:
I just realized I never implemented the cutoff in the actually lookups 🤦 .

My intent was just to min the input density against the maximum. It looks like `std::numerical_limits::infinity` needs a special `max()` function. Is there a way to avoid this?

Reply:
I think `std::min(rho, std::numerical_limits::infinity());` does what you want? Am I missing something?

Reply:
Yeah I don't know what I was thinking... I'll implement it